[OpInfo] Added ReductionOpInfo subclass of OpInfo and ported sum test

heitorschueroff opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63978
* #62900
* #62899
* __->__ #62737

ReductionOpInfo is a specialization of OpInfo for reduction operators. For now, it is designed to work with reductions that return a single tensor and that reduce all elements along one or more dimensions to a single value. In particular this excludes operators such as `max` and `min` that return multiple tensors and `quantile` that can return multiple values.

fixes https://github.com/pytorch/pytorch/issues/49746

Differential Revision: [D30406568](https://our.internmc.facebook.com/intern/diff/D30406568)

heitorschueroff committed:
[OpInfo] Added ReductionOpInfo subclass of OpInfo

[ghstack-poisoned]

heitorschueroff committed:
Update on "[OpInfo] Added ReductionOpInfo subclass of OpInfo"




[ghstack-poisoned]

heitorschueroff committed:
Update on "[OpInfo] Added ReductionOpInfo subclass of OpInfo"




[ghstack-poisoned]

heitorschueroff committed:
Update on "[OpInfo] Added ReductionOpInfo subclass of OpInfo"




[ghstack-poisoned]

heitorschueroff committed:
Update on "[OpInfo] Added ReductionOpInfo subclass of OpInfo"




[ghstack-poisoned]

heitorschueroff committed:
Update on "[OpInfo] Added ReductionOpInfo subclass of OpInfo"




[ghstack-poisoned]

heitorschueroff committed:
Update on "[OpInfo] Added ReductionOpInfo subclass of OpInfo"




[ghstack-poisoned]

heitorschueroff committed:
Update on "[OpInfo] Added ReductionOpInfo subclass of OpInfo"




[ghstack-poisoned]

heitorschueroff committed:
Update on "[OpInfo] Added ReductionOpInfo subclass of OpInfo and ported sum test"




[ghstack-poisoned]

heitorschueroff committed:
Update on "[OpInfo] Added ReductionOpInfo subclass of OpInfo and ported sum test"




[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -42,6 +42,15 @@
     import scipy.special
 
 
+# Reasonable testing sizes for dimensions
```
Moving these here is a good idea

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -750,9 +758,165 @@ def default_test_dtypes(self, device_type):
+# with ndim dims appropriate for testing. If supports_multiple_dims
+# is True (default) then dim kwarg can be a list of dims.
+def _generate_reduction_kwargs(ndim, supports_multiple_dims=True):
+    for keepdim in [True, False]:
```
Should this also yield a no kwargs case?

mruberry replied:
OK it's added explicitly below. That seems fine

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -750,9 +758,165 @@ def default_test_dtypes(self, device_type):
+
+
+# Wraps sample_inputs_reduction function to provide the additional supports_multiple_dims args
+def sample_inputs_reduction_wrapper(supports_multiple_dims):
```
A possibly simpler option would be to:

```
def sample_inputs_reductions(op_info, device, dtype, requires_grad, *, supports_multiple_dims, **kwargs):
```

And then the sample inputs for an OpInfo would partial the `supports_multiple_dims` kwarg like the div OpInfos partial their sample input func.

heitorschueroff replied:
I didn't know about `partial` but that is a great suggestion, thank you!

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -750,9 +758,165 @@ def default_test_dtypes(self, device_type):
+        inputs = []
+
+        for t in _generate_reduction_inputs(device, dtype, requires_grad):
+            # Add case without dim and keepdim kwargs
```
This special case not needed if _generate_reduction_kwargs returned `{}`, too

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -750,9 +758,165 @@ def default_test_dtypes(self, device_type):
+
+# NOTE [Reductions]:
+#
+# For the purpose of testing, we relax the definition of a reduction operator
```
Great comment. We can also revisit this decision later since all this is internal

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -750,9 +758,165 @@ def default_test_dtypes(self, device_type):
+        **kwargs,
+    ):
+        # These are mutually exclusive options
+        assert not (promotes_int_to_float and promotes_int_to_int64)
```
Nice checks

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -750,9 +758,165 @@ def default_test_dtypes(self, device_type):
+
+        assert nan_policy in ('propagate', 'omit')
+
+        # ReductionOpInfo tests generate their own input tensors and the dim
```
Nice comment

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -750,9 +758,165 @@ def default_test_dtypes(self, device_type):
+        # or specify `sample_inputs_func` to provide their own sample inputs for
+        # the OpInfo base class. However, the test_reductions.py file will still
+        # generate its own inputs.
+        def sample_inputs_func(*args, **kwargs):
```
I understand the comment but I'm not sure I understand why the structure of this function is necessary -- should the default sample inputs func just be a partial of one of the above funcs? Why the kwarg surgery here?

heitorschueroff replied:
This is necessary to augment the sample inputs from `sample_inputs_reduction` with the args and kwargs from `generate_args_kwargs`.

mruberry replied:
I guess I can buy that. An alternative design might have had `sample_inputs_reduction` accept the `generate_args_kwargs` function as an input and perform this logic itself

heitorschueroff replied:
Yes, that need actually came up later when porting other ops so I might do that. Thanks for the suggestion.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -750,9 +758,165 @@ def default_test_dtypes(self, device_type):
+            return result
+
+        # Override OpInfo defaults and call base class __init__
+        kwargs.setdefault('inplace_variant', None)
```
Neat

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -8350,6 +8459,20 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
 ]
 
+# Reduction operator database
+reduction_op_db: List[ReductionOpInfo] = [
```
Add it to the main ops db but then create a filtered list like unary_ufuncs does for consistency

mruberry reviewed (commented):
Overall looks very good, @heitorschueroff; just a few inline questions/suggestions for your review.

heitorschueroff commented on test/test_reductions.py:
```diff
@@ -55,6 +57,21 @@ def _rand_shape(dim, min_size, max_size):
+    def test_dim_default(self, device, dtype, op: ReductionOpInfo):
+        """Tests that the default behavior is to reduce all dimensions."""
+        t = make_tensor((2, 3), device, dtype)
+        args, kwargs = next(op.generate_args_kwargs(t))
```
the test generates dim and keepdim kwargs and passed them along with the extra args and kwargs generated by the op. In this case, we do not pass dim and keepdim to op to test the default behavior. However, I'm not checking that `generate_args_kwargs` doesn't actually generate dim and keepdim and if it did then this test would probably fail.

mruberry replied:
Right, but isn't that worth addressing?

mruberry replied:
OK my mistake I was conflating two functions (there are a lot of functions). I guess it's reasonable to assume `generate_args_kwargs` won't generate `dim` or `keepdim`.

heitorschueroff committed:
Update on "[OpInfo] Added ReductionOpInfo subclass of OpInfo and ported sum test"




[ghstack-poisoned]

heitorschueroff committed:
Update on "[OpInfo] Added ReductionOpInfo subclass of OpInfo and ported sum test"




[ghstack-poisoned]

heitorschueroff committed:
Update on "[OpInfo] Added ReductionOpInfo subclass of OpInfo and ported sum test"




[ghstack-poisoned]

heitorschueroff committed:
Update on "[OpInfo] Added ReductionOpInfo subclass of OpInfo and ported sum test"




[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -769,9 +777,163 @@ def default_test_dtypes(self, device_type):
+    yield {}
+
+    # Test reducing inner and outer most dimensions
+    yield {'dim': 0, 'keepdim': True}
```
Like if this were moved above the first yield wouldn't the test above fail?

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -8350,6 +8459,20 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
 
+# Reduction operator database
+reduction_op_db: List[ReductionOpInfo] = [
+    ReductionOpInfo(
```
Add this alphabetically (as best you can since the list isn't alphabetical) instead of at the end of the sequence to help avoid/prevent merge conflicts

heitorschueroff replied:
That was my goals with putting it into a reductions_op_db initially, but I guess we should just do a refactor of common_methods_invocations at some point. I'll add them alphabetically.

mruberry reviewed (approved):
A few small suggestions inline for you to consider, but overall things look good. Hurray ReductionInfos!

heitorschueroff committed:
Update on "[OpInfo] Added ReductionOpInfo subclass of OpInfo and ported sum test"


ReductionOpInfo is a specialization of OpInfo for reduction operators. For now, it is designed to work with reductions that return a single tensor and that reduce all elements along one or more dimensions to a single value. In particular this excludes operators such as `max` and `min` that return multiple tensors and `quantile` that can return multiple values.

fixes https://github.com/pytorch/pytorch/issues/49746

[ghstack-poisoned]

heitorschueroff commented:
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30406568).

heitorschueroff committed:
Update on "[OpInfo] Added ReductionOpInfo subclass of OpInfo and ported sum test"


ReductionOpInfo is a specialization of OpInfo for reduction operators. For now, it is designed to work with reductions that return a single tensor and that reduce all elements along one or more dimensions to a single value. In particular this excludes operators such as `max` and `min` that return multiple tensors and `quantile` that can return multiple values.

fixes https://github.com/pytorch/pytorch/issues/49746

Differential Revision: [D30406568](https://our.internmc.facebook.com/intern/diff/D30406568)

[ghstack-poisoned]

heitorschueroff commented:
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30406568).

heitorschueroff committed:
Update on "[OpInfo] Added ReductionOpInfo subclass of OpInfo and ported sum test"


ReductionOpInfo is a specialization of OpInfo for reduction operators. For now, it is designed to work with reductions that return a single tensor and that reduce all elements along one or more dimensions to a single value. In particular this excludes operators such as `max` and `min` that return multiple tensors and `quantile` that can return multiple values.

fixes https://github.com/pytorch/pytorch/issues/49746

Differential Revision: [D30406568](https://our.internmc.facebook.com/intern/diff/D30406568)

[ghstack-poisoned]

heitorschueroff commented:
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30406568).

heitorschueroff committed:
Update on "[OpInfo] Added ReductionOpInfo subclass of OpInfo and ported sum test"


ReductionOpInfo is a specialization of OpInfo for reduction operators. For now, it is designed to work with reductions that return a single tensor and that reduce all elements along one or more dimensions to a single value. In particular this excludes operators such as `max` and `min` that return multiple tensors and `quantile` that can return multiple values.

fixes https://github.com/pytorch/pytorch/issues/49746

Differential Revision: [D30406568](https://our.internmc.facebook.com/intern/diff/D30406568)

[ghstack-poisoned]

heitorschueroff commented:
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30406568).

heitorschueroff committed:
Update on "[OpInfo] Added ReductionOpInfo subclass of OpInfo and ported sum test"


ReductionOpInfo is a specialization of OpInfo for reduction operators. For now, it is designed to work with reductions that return a single tensor and that reduce all elements along one or more dimensions to a single value. In particular this excludes operators such as `max` and `min` that return multiple tensors and `quantile` that can return multiple values.

fixes https://github.com/pytorch/pytorch/issues/49746

Differential Revision: [D30406568](https://our.internmc.facebook.com/intern/diff/D30406568)

[ghstack-poisoned]

heitorschueroff commented:
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30406568).

heitorschueroff committed:
Update on "[OpInfo] Added ReductionOpInfo subclass of OpInfo and ported sum test"


ReductionOpInfo is a specialization of OpInfo for reduction operators. For now, it is designed to work with reductions that return a single tensor and that reduce all elements along one or more dimensions to a single value. In particular this excludes operators such as `max` and `min` that return multiple tensors and `quantile` that can return multiple values.

fixes https://github.com/pytorch/pytorch/issues/49746

Differential Revision: [D30406568](https://our.internmc.facebook.com/intern/diff/D30406568)

[ghstack-poisoned]

heitorschueroff committed:
Update on "[OpInfo] Added ReductionOpInfo subclass of OpInfo and ported sum test"


ReductionOpInfo is a specialization of OpInfo for reduction operators. For now, it is designed to work with reductions that return a single tensor and that reduce all elements along one or more dimensions to a single value. In particular this excludes operators such as `max` and `min` that return multiple tensors and `quantile` that can return multiple values.

fixes https://github.com/pytorch/pytorch/issues/49746

Differential Revision: [D30406568](https://our.internmc.facebook.com/intern/diff/D30406568)

[ghstack-poisoned]
